### PR TITLE
Implement CSV export for Matrix vis

### PR DESCRIPTION
**Pull request by axelboc**
Part of #35 

This is the simplest step I could think of to make some data exportable from H5Web. The Matrix vis now shows a button to download the current slice as CSV.

image

image

### CODE REVIEW

File: src/h5web/toolbar/controls/Btn.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  label: string;
+  icon?: IconType;
+  iconOnly?: boolean;
+  small?: boolean;
+  raised?: boolean;
+  onClick: () => void;
+  disabled?: boolean;
+}
+
+function Btn(props: Props) {
```
Comment (axelboc) [inline src/h5web/toolbar/controls/Btn.tsx:15]: This is unrelated (see commits) -- I'm just abstracting a `Btn` component out of `ToggleBtn` and I expose it to @h5web/lib.

File: src/h5web/toolbar/controls/DownloadBtn.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  label: string;
+  icon?: IconType;
+  iconOnly?: boolean;
+  // If specified, browser will prompt user to download the file instead of opening a new tab
+  filename?: string;
+  // Invoked on click; must return URL to set as `href`, or `false` to do nothing
+  getDownloadUrl: () => string | false;
+}
+
+function DownloadBtn(props: Props) {
```
Comment (axelboc) [inline src/h5web/toolbar/controls/DownloadBtn.tsx:15]: The CSV export button behaves the same way as the `SnapshotBtn`, so I thought I'd make an abstraction as well.

File: src/h5web/toolbar/MatrixToolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import { FiDownload } from 'react-icons/fi';
+import { useMatrixConfig } from '../vis-packs/core/matrix/config';
+import DownloadBtn from './controls/DownloadBtn';
+import Toolbar from './Toolbar';
+
+function MatrixToolbar() {
+  const currentSlice = useMatrixConfig((state) => state.currentSlice);
```
Comment (axelboc) [inline src/h5web/toolbar/MatrixToolbar.tsx:7]: The toolbar needs access to the `ndarray` of the current slice. That's why I need a zustand state, which I set in `MappedMatrixVis`, via a `useEffect` (just like `dataDomain`).

**Review by loichuder — CHANGES_REQUESTED** [review]
The export does not work for 1D data so you can either change the `throw` condition to reflect this or change the algo to handle 1D data.

File: src/h5web/toolbar/MatrixToolbar.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  if (currentSlice && currentSlice.shape.length > 2) {
+    throw new Error('Expected current slice to have at most two dimensions');
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/toolbar/MatrixToolbar.tsx:10]: Silly me... 😓

File: src/h5web/toolbar/MatrixToolbar.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        <DownloadBtn
+          icon={FiDownload}
+          label="CSV"
+          filename="export.csv"
+          getDownloadUrl={() => {
+            let data = '';
+            for (let i = 0; i < currentSlice.shape[0]; i++) {
+              let line = '';
+              for (let j = 0; j < currentSlice.shape[1]; j++) {
+                line += `${currentSlice.get(i, j).toString()},`; // complex numbers are stringifyied as two values
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/toolbar/MatrixToolbar.tsx]: Yeah not sure. Two values is simpler as we don't have to check whether each value is an array or not (we don't currently have access to the dataset's type in the toolbar) -- `toString()` just outputs `a,b` when the value is of type `[number, number]`. The other advantage, I find, is for users: they can do phase/amplitude/... computations on the data without having to parse values of the form `a+b*i`.

The downside is that users end up with twice as many columns as they may expect and we don't make that clear anywhere. However, I think it's pretty easy to understand by just comparing a few values between the file and H5Web.

File: src/h5web/vis-packs/core/matrix/utils.ts [hunk 6]
```diff
@@ -21,3 +23,29 @@ export function getFormatter(
...
+    return csv;
+  }
+
+  if (slice.shape.length === 2) {
+    for (let i = 0; i < slice.shape[0]; i++) {
+      let line = '';
+      for (let j = 0; j < slice.shape[1]; j++) {
+        line += `${slice.get(i, j).toString()},`; // complex numbers are stringifyied as two values
+      }
+      csv += `${line.replace(/,$/u, '')}\n`;
```
Comment (loichuder) [inline src/h5web/vis-packs/core/matrix/utils.ts:44]:
```suggestion
      csv += `${line.slice(0, -1)}\n`;
```
or
```suggestion
      csv += line.replace(/,$/u, '\n');
```

File: src/h5web/vis-packs/core/matrix/utils.ts [hunk 7]
```diff
@@ -21,3 +23,29 @@ export function getFormatter(
...
+
+export function sliceToCsv(slice: NdArray<Primitive<PrintableType>>): string {
+  let csv = '';
+
+  if (slice.shape.length === 1) {
+    for (let i = 0; i < slice.shape[0]; i++) {
+      csv += `${slice.get(i).toString()}\n`; // complex numbers are stringifyied as two values
+    }
+
+    return csv;
```
Comment (loichuder) [inline src/h5web/vis-packs/core/matrix/utils.ts:35]: Alternatively, you can make use of `slice.data` for 1D slices:
```suggestion
    return slice.data.reduce((acc, val) => acc += `${val.toString()}\n`;
```

and keep `let csv = '';` inside `if (slice.shape.length === 2)` block.

**Review by loichuder — APPROVED** [review]

File: src/h5web/vis-packs/core/matrix/utils.ts (hunk 7 above)
Comment (axelboc) [reply src/h5web/vis-packs/core/matrix/utils.ts:35]: TypeScript doesn't like it, even with a type cast on `slice.data` 😢 -- maybe once the typing improvements of `ndarray` are merged...

File: src/h5web/vis-packs/core/matrix/utils.ts (hunk 6 above)
Comment (axelboc) [reply src/h5web/vis-packs/core/matrix/utils.ts:44]: Ha, clever 😄 -- I'll use the second one, as I like the self-documenting aspect of the regex.
